modules/pipeline/providers/cron/cron.service.go

```diff
@@ -229,3 +242,177 @@ func (s *provider) CronUpdate(ctx context.Context, req *pb.CronUpdateRequest) (*
+	}
+	v2Exist, err := s.dbClient.DB().Get(queryV2)
+	if err != nil {
+		return err
+	}
+	if v2Exist {
+		new_.ID = queryV2.ID
+		new_.Enable = queryV2.Enable
+		err := s.dbClient.UpdatePipelineCron(new_.ID, new_, ops...)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err = s.dbClient.CreatePipelineCron(new_, ops...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *provider) CronDisable(ctx context.Context, req *pb.CronDisableRequest) (*pb.CronDisableResponse, error) {
+
+	cron := pbCronToDBCron(req.Cron)
+
+	txSession := s.dbClient.NewSession()
+	defer txSession.Close()
+	if err := txSession.Begin(); err != nil {
+		return nil, err
+	}
+	err := func() error {
+
+		defer func() {
+			if err := s.Daemon.AddIntoPipelineCrond(cron.ID); err != nil {
+				s.Log.Errorf("[alert] add crond failed, err: %v", err)
+			}
+		}()
+
+		var disable = false
+		var updateCron = &db.PipelineCron{}
+		var columns = []string{spec.PipelineCronCronExpr, spec.PipelineCronEnable}
+		var err error
+
+		queryV1 := &db.PipelineCron{
+			ApplicationID:   cron.ApplicationID,
+			Branch:          cron.Branch,
+			PipelineYmlName: cron.PipelineYmlName,
+		}
+		v1Exist, err := txSession.Get(queryV1)
+		if err != nil {
+			return err
+		}
+		if queryV1.Extra.Version == "v2" {
+			v1Exist = false
+		}
+		if v1Exist {
+			updateCron.Enable = &disable
+			updateCron.ID = queryV1.ID
+			updateCron.CronExpr = cron.CronExpr
+			err := s.dbClient.UpdatePipelineCronWillUseDefault(updateCron.ID, updateCron, columns, mysqlxorm.WithSession(txSession))
+			if err != nil {
+				return err
+			}
+			req.Cron.ID = updateCron.ID
+
+			return nil
+		}
+
+		queryV2 := &spec.PipelineCron{
+			PipelineSource:  cron.PipelineSource,
+			PipelineYmlName: cron.PipelineYmlName,
+		}
+		v2Exist, err := txSession.Get(queryV2)
+		if err != nil {
+			return err
+		}
+		if v2Exist {
+			updateCron.Enable = &disable
+			updateCron.ID = queryV2.ID
+			updateCron.CronExpr = cron.CronExpr
+			err := s.dbClient.UpdatePipelineCronWillUseDefault(updateCron.ID, updateCron, columns, mysqlxorm.WithSession(txSession))
+			if err != nil {
+				return err
+			}
+			req.Cron.ID = updateCron.ID
+			return nil
+		}
+
+		return nil
+	}()
+
+	if err != nil {
+		if rbErr := txSession.Rollback(); rbErr != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+	if cmErr := txSession.Commit(); cmErr != nil {
+		return nil, cmErr
+	}
+
+	return &pb.CronDisableResponse{
+		Data: req.Cron,
+	}, nil
+}
+
+func pbCronToDBCron(pbCron *common.Cron) *db.PipelineCron {
+	cron := db.PipelineCron{
+		ID:                   pbCron.ID,
+		ApplicationID:        pbCron.ApplicationID,
+		Branch:               pbCron.Branch,
+		PipelineSource:       apistructs.PipelineSource(pbCron.PipelineSource),
+		PipelineYmlName:      pbCron.PipelineYmlName,
+		CronExpr:             pbCron.CronExpr,
+		Enable:               &[]bool{pbCron.Enable.Value}[0],
+		PipelineDefinitionID: pbCron.PipelineDefinitionID,
```

Comment:
where is version field?

I think this method is not good, it's very easy to lose some fields, and no one knows to change here.

maybe JSON serialization is enough?